packages/context/src/context.ts

```diff
@@ -353,16 +405,25 @@ export class Context {
    *   on how the binding was configured.
    * @internal
    */
-  getValueOrPromise(
+  getValueOrPromise<T>(
     keyWithPath: string,
     optionsOrSession?: ResolutionOptions | ResolutionSession,
-  ): ValueOrPromise<BoundValue> {
+  ): T | PromiseLike<T> | undefined {
```

Comment:
Should it be `ValueOrPromise<T | undefined>`? If the `optional` is `true`, can a value provider return `undefined`?

Reply:
This is a good question. Right now, the code in `getValueOrPromise` returns `undefined` only when the binding was not found. A provider cannot return `undefined` unless it's declared as `implements Provider</*...*/ | undefined>`, which is actually a valid usage we need to consider!

Another source of `undefined` is `getDeepProperty`, which can return `undefined` when the deep property does not exist. 

**I need to take another look at both issues pointed above.**

I think that `ctx.get` should throw when a deep property does not exist, for consistency with the case when the binding itself was not found. What do you think?

Would you mind if I restrict the changes in this patch to changing the return value of `getValueOrPromise` to `ValueOrPromise<T  | undefined>`, and then look how to further improve type checks (and possibly throw when a deep property was not found) in a follow-up pull request?